src/main/java/br/com/caelum/vraptor/jpa/EntityManagerFactoryCreator.java

```diff
@@ -33,15 +37,52 @@
  * @author Otávio Garcia
  */
 public class EntityManagerFactoryCreator {
-	
-	@Inject 
+	@Inject
 	private Environment environment;
+	@Inject
+	private Properties propertiesOfJPAConnection;
 
+	
+	/**
+	 * Produces EntityManagerFactory.  For default  is created an connect 
+	 * based on configurations of the persistence.xml. 
+	 * However, if the database properties are unset in persistence.xml,  
+	 * the connection will created programmatically by propertiesOfJPAConnection,
+	 * overriding the default settings.
+	 * 
+	 * @return the EntityManagerFactory that will create all Entity managers
+	 * @throws IOException 
+	 * @throws FileNotFoundException 
+	 */
 	@ApplicationScoped
 	@Produces
-	public EntityManagerFactory getEntityManagerFactory() {
+	public EntityManagerFactory getEntityManagerFactory(){
+		EntityManagerFactory factory = null;
 		String persistenceUnit = environment.get("br.com.caelum.vraptor.jpa.persistenceunit", "default");
-		return Persistence.createEntityManagerFactory(persistenceUnit);
+		
+		if(!propertiesOfJPAConnection.isEmpty() && !hasXMLConfiguration(persistenceUnit)) {
```

Comment:
You are creating `EntityManagerFactory` twice. Since `EntityManagerFactory` creation are too expensive, we can't create twice. The 1st creation are inside this if (line 79), to check properties, and another there inside this condition (lines 64 or 68).